File: LeetCode/0209. Minimum Size Subarray Sum/Solution.py
class Solution:
    def minSubArrayLen(self, s, nums):
        """
        :type s: int
        :type nums: List[int]
        :rtype: int
        ------------------------------
        15 / 15 test cases passed.
        Status: Accepted
        Runtime: 72 ms
        """
        l, sum, res = 0, 0, len(nums) + 1
        for r in range(len(nums)):
            sum += nums[r]
            while l <= r and sum >= s:
                res = min(res, r - l + 1)
                sum -= nums[l]
                l += 1

        return res if res != len(nums) + 1 else 0


if __name__ == '__main__':
    print(Solution().minSubArrayLen(15, [1, 2, 3, 4, 5]))
    print(Solution().minSubArrayLen(7, [2, 3, 1, 2, 4, 3]))

File: LeetCode/0209. Minimum Size Subarray Sum/Solution.kt
package io.innofang

/**
 * Created by Inno Fang on 2018/2/7.
 */

/**
 * 15 / 15 test cases passed.
 * Status: Accepted
 * Runtime: 422 ms
 */
class Solution {
    fun minSubArrayLen(s: Int, nums: IntArray): Int {
        var l = 0
        var r = -1
        var sum = 0
        var res = nums.size + 1

        while (l < nums.size) {
            if (r + 1 < nums.size && sum < s) sum += nums[++r]
            else sum -= nums[l++]

            if (sum >= s) res = minOf(r - l + 1, res)
        }
        return if (res != nums.size + 1) res else 0
    }
}

fun main(args: Array<String>) {
    Solution().minSubArrayLen(7, intArrayOf(2,3,1,2,4,3)).let(::println)
    Solution().minSubArrayLen(15, intArrayOf(1,2,3,4,5)).let(::println)
}